torch/csrc/cuda/AutoGPU.cpp

```diff
@@ -59,12 +60,14 @@ static int getObjDevice(PyObject *args, PyObject *self) {
   return -1;
 }
 
-THCPAutoGPU::THCPAutoGPU(int device_id) : AutoGPU(device_id) {}
-
 THCPAutoGPU::THCPAutoGPU(PyObject *args, PyObject *self)
   : AutoGPU(getObjDevice(args, self)) {
 }
 
 void THCPAutoGPU::setObjDevice(PyObject *obj) {
   setDevice(getObjDevice(obj));
 }
+#endif
+
+THCPAutoGPU::THCPAutoGPU(int device_id) : AutoGPU(device_id) {}
```

Comment:
can't you just move the constructor implementation to the header and compile this file only conditionally?